LION/models/PnP/gs_drunet.py

```diff
@@ -0,0 +1,76 @@
+# This file is part of LION library
+# License : GPL-3
+#
+# Author: Ferdia Sherry
+# Modifications: -
+# =============================================================================
+
+from .drunet import DRUNet
+from LION.models.LIONmodel import LIONmodel, LIONParameter
+
+from typing import Optional
+import torch
+
+
+class GSDRUNet(LIONmodel):
+    def __init__(self, model_parameters: LIONParameter = None):
+        if model_parameters is None:
+            model_parameters = GSDRUNet.default_parameters()
+        super().__init__(model_parameters)
+        self._backbone_model = DRUNet(model_parameters)
+
+    @staticmethod
+    def default_parameters():
+        return LIONParameter(
+            in_channels=1,
+            out_channels=1,
+            int_channels=32,
+            kernel_size=(3, 3),
+            n_blocks=2,
+            use_noise_level=False,
+            bias_free=False,
+            act="elu",
+            enforce_positivity=False,
+        )
+
+    @staticmethod
+    def cite(cite_format="MLA"):
+        if cite_format == "MLA":
+            print("Hurault, Samuel, Arthur, Leclaire, Nicolas, Papadakis.")
+            print('"Gradient Step Denoiser for Convergent Plug-and-Play."')
+            print(
+                "\x1b[3mInternational Conference on Learning Representations 2022\x1b[0m."
+            )
+
+        elif cite_format == "bib":
+            print("@inproceedings{hurault2022gradient,")
+            print("title = {Gradient Step Denoiser for Convergent Plug-and-Play},")
+            print(
+                "booktitle = {International {{Conference}} on {{Learning Representations}}},"
+            )
+            print(
+                "author = {Hurault, Samuel and Leclaire, Arthur and Papadakis, Nicolas},"
+            )
+            print("year = {2022}")
+            print("}")
+        else:
+            raise ValueError(
+                f'`cite_format` "{cite_format}" is not understood, only "MLA" and "bib" are supported'
+            )
+
+    def obj_grad(self, x, noise_level: Optional[float] = None):
+        x = x.detach()
+        x.requires_grad = True
```

Comment:
Can you explain this? So you remove it from the computational graph, but then you start a new computaional graph by requiring grad?

Given that `forward` itself has no loops, how is this doing anything? Or is it to do with something about the way you trained?

Reply:
This gives you a new tensor with the same underlying data but which doesn't interfere with previous things done on `x`. If I removed line 62, I would be messing with the `x` that the user has input (undesirable), whereas now after line 62 `x` refers to a new tensor so I can do with it what I want (except modify it). I could have just as well said `y = x.detach()` in line 62 and then used `y` for the rest of it. The screenshot shows what I mean:
<img width="675" alt="Screenshot 2024-07-15 at 23 21 18" src="https://github.com/user-attachments/assets/2321865e-ef12-444b-adcf-06de135ac189">

Reply:
Sorry,  I am still confused :D. Yes, I do understand what it does in principle, but when we make ML models, we don't need to detach the input of the model, right? the computational graph understands that its the input of the model and starts the graph there. given that `obj_grad` is just used in `forward`, why is this line needed here?

Reply:
You need to do this because we do a backward pass inside `obj_grad`. Otherwise you mess up the input `x`